### PR TITLE
[PLUGIN-632] - Added: ServiceAccountType as JSON, unit tests, validations

### PR DESCRIPTION
**Pull request by flakrimjusufi**
Make Google drive/sheet plugins production ready
JIRA Ticket: https://cdap.atlassian.net/browse/PLUGIN-632

### CODE REVIEW

**Comment by jasir99** [comment]
lgtm

File: src/main/java/io/cdap/plugin/google/drive/source/GoogleDriveInputFormat.java [hunk 1]
```diff
@@ -18,6 +18,9 @@
 
 import com.github.rholder.retry.RetryException;
 import com.google.api.services.drive.model.File;
+import com.google.gson.JsonObject;
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/google/drive/source/GoogleDriveInputFormat.java]: Why are these imports added?

File: src/main/java/io/cdap/plugin/google/drive/source/SchemaBuilder.java [hunk 2]
```diff
@@ -27,7 +27,7 @@
  * Util class for building pipeline schema.
  */
 public class SchemaBuilder {
-  public static final String SCHEMA_ROOT_RECORD_NAME = "FileFromFolder";
+  public static final String SCHEMA_ROOT_RECORD_NAME = "etlSchemaBody";
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/google/drive/source/SchemaBuilder.java]: why was this change needed?

File: src/main/java/io/cdap/plugin/google/common/GoogleDriveClient.java [hunk 3]
```diff
@@ -98,7 +105,7 @@ protected Credential getCredentials() throws IOException {
   }
 
   protected List<String> getRequiredScopes() {
-    return Collections.singletonList(DriveScopes.DRIVE_READONLY);
+    return Collections.singletonList(DriveScopes.DRIVE);
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/google/common/GoogleDriveClient.java]: why did we need to make this change?

File: src/main/java/io/cdap/plugin/google/common/GoogleAuthBaseConfig.java [hunk 4]
```diff
@@ -158,25 +176,30 @@ private boolean validateOAuth2Properties(FailureCollector collector) {
       & checkPropertyIsSet(collector, refreshToken, REFRESH_TOKEN, REFRESH_TOKEN_LABEL);
   }
 
-  private boolean validateAccountFilePath(FailureCollector collector) {
-    if (!containsMacro(ACCOUNT_FILE_PATH)) {
+  private boolean validateServiceAccount(FailureCollector collector) {
+    if (isServiceAccountFilePath() != null && isServiceAccountFilePath()) {
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/google/common/GoogleAuthBaseConfig.java]: store the return value in a variable so you don't have to call `isServiceAccountFilePath()` twice. Same thing for `isServiceAccountJson()` below.

**Review by rmstar — COMMENTED** [review]
Please test source and sink plugins in the following scenarios:
1. In CDF with file path set to auto-detect.
2. In CDF, specify file path
3. In CDF, specify Service Account JSON
4. In CDAP, specify file path
5. In CDAP, specify Service Account JSON
6. Test with service account type, file path/JSON set as macro param.

File: src/main/java/io/cdap/plugin/google/drive/source/GoogleDriveInputFormat.java (hunk 1 above)
Comment (flakrimjusufi) [reply src/main/java/io/cdap/plugin/google/drive/source/GoogleDriveInputFormat.java]: These imports were left behind after we did some additional tests. Thanks for letting me know. I removed them.

File: src/main/java/io/cdap/plugin/google/drive/source/SchemaBuilder.java (hunk 2 above)
Comment (flakrimjusufi) [reply src/main/java/io/cdap/plugin/google/drive/source/SchemaBuilder.java]: The schema was not being updated in _Google Drive Source_ and initially we thought that this was the case, but then we found out that the issue was related with widget, not with `SCHEMA_ROOT_RECORD_NAME`. Reverted to its initial name in the updated PR.

File: src/main/java/io/cdap/plugin/google/common/GoogleDriveClient.java (hunk 3 above)
Comment (flakrimjusufi) [reply src/main/java/io/cdap/plugin/google/common/GoogleDriveClient.java]: This was also as part of the testings that we have done. We were dealing with this error message when testing the pipeline:

`404 Not Found { "code" : 404, "errors" : [ { "domain" : "global", "location" : "fileId", "locationType" : "parameter", "message" : "File not found: 1ZqAr0q6gITvjqvfOJqkSIxd-usTNhvJy.", "reason" : "notFound" } ], "message" : "File not found: 1ZqAr0q6gITvjqvfOJqkSIxd-usTNhvJy." }. Please check the system logs for more details.
`

...and checking the documentation of Google Drive API for 404 error type, stating that:

To fix this error:

```
1. Inform the user they don't have read access to the file or that the file doesn't exist.
2. Instruct the user to ask the owner for permission to the file.
```
...we changed the SCOPE from `DRIVE_READONLY` to `DRIVE`.

But afterwards, when we verified that in fact we were not in the right path when testing the pipelines and DRIVE_READONLY was enough for read/write, we changed back to its initial SCOPE. 

In the last commit, we reverted back the scope to `DRIVE_READONLY`.

File: src/main/java/io/cdap/plugin/google/common/GoogleAuthBaseConfig.java (hunk 4 above)
Comment (flakrimjusufi) [reply src/main/java/io/cdap/plugin/google/common/GoogleAuthBaseConfig.java]: PR updated.

**Comment by flakrimjusufi** [comment]
> Please test source and sink plugins in the following scenarios:
> 
> 1. In CDF with file path set to auto-detect.
> 2. In CDF, specify file path
> 3. In CDF, specify Service Account JSON
> 4. In CDAP, specify file path
> 5. In CDAP, specify Service Account JSON
> 6. Test with service account type, file path/JSON set as macro param.

1. I'm unable to test with file path set to auto-detect in CDF because my keys have insufficient permissions. To be more specific, I've used `flakrim@cirus.co` to test the plugin since I don't have access to Google Drive with my partner account `flakrim@88547414.corp-partner.google.com`. In CDF, we only have access with partnership accounts. 

2. Is there a way to test the pipelines in CDF while specifying the file path? As far as I know, in CDF we can only test with Service Account as JSON. Please correct me if my assumption is wrong. 

3. Test passed successfully. 

4. Test passed successfully. 

5. Test passed successfully. 

6. There's an issue when you test with MACRO. We were getting this error message:

`Java.util.concurrent.ExecutionException: com.google.gson.JsonSyntaxException: com.google.gson.stream.MalformedJsonException: Unterminated object at line 1 column 5965
	at com.google.common.util.concurrent.AbstractFuture$Sync.getValue(AbstractFuture.java:294) ~[com.google.guava.guava-13.0.1.jar:na]
	at com.google.common.util.concurrent.AbstractFuture$Sync.get(AbstractFuture.java:281) ~[com.google.guava.guava-13.0.1.jar:na]
	at com.google.common.util.concurrent.AbstractFuture.get(AbstractFuture.java:116) ~[com.google.guava.guava-13.0.1.jar:na]
`
That's because of this block of code: 

https://github.com/data-integrations/google-drive/blob/a9be71186e7aae818c5b5c78a5d33e254c1f5368/src/main/java/io/cdap/plugin/google/drive/source/GoogleDriveInputFormatProvider.java#L40-L43

After debugging this issue, we found out that the entire config was serialized. 

Taking a consideration that we have to do a deserialization of the entire config and this is time consuming (after this fix, the entire functionality of source and sink plugin needs to be tested), we removed Macros at this moment of time. We were told that the requirements were prioritized and the fix needs to be delivered as soon as possible.

However, we can proceed with the deserialization of config and we can include Macros in this PR if needed.

**Comment by CuriousVini** [comment]
@flakrimjusufi
Do you know what permissions are missing from your account?

**Comment by CuriousVini** [comment]
Also are you noticing permissions issues at pipeline runtime? If so, could you please make sure dataproc service account (GCE default service account) has permissions to access drive directories/files.

**Comment by rmstar** [comment]
> Taking a consideration that we have to do a deserialization of the entire config and this is time consuming (after this fix, the entire functionality of source and sink plugin needs to be tested), we removed Macros at this moment of time. We were told that the requirements were prioritized and the fix needs to be delivered as soon as possible.

Macros are supported, so we should make sure they work correctly. Please fix it as part of this PR.

File: src/main/java/io/cdap/plugin/google/common/GoogleDriveClient.java (hunk 3 above)
Comment (rmstar) [reply src/main/java/io/cdap/plugin/google/common/GoogleDriveClient.java]: With SCOPE set to `DRIVE` does everything work correctly?

Comment (flakrimjusufi) [reply src/main/java/io/cdap/plugin/google/common/GoogleDriveClient.java]: Actually, everything also works correctly with SCOPE set to `DRIVE_READONLY`, that's why we reverted back the code.

**Comment by flakrimjusufi** [comment]
> @flakrimjusufi
> Do you know what permissions are missing from your account?

With our partnership accounts, we don't have access in any of gmail, google drive, etc. So, to test Google Drive Source/Sink in data proc, I need one of the followings:  

- My account in cirus (which I've used to create a project and work with Google Drive API) to have access in data proc. 
- My partnership account to have access in Google Drive API. 

Permission issues are showing up in validation when file path is set to `auto-detect` in datafusion, I can't run the pipeline.

**Comment by flakrimjusufi** [comment]
> > Taking a consideration that we have to do a deserialization of the entire config and this is time consuming (after this fix, the entire functionality of source and sink plugin needs to be tested), we removed Macros at this moment of time. We were told that the requirements were prioritized and the fix needs to be delivered as soon as possible.
> 
> Macros are supported, so we should make sure they work correctly. Please fix it as part of this PR.

@rmstar 

In the last update, we deserialized the config and added Macros.

File: src/main/java/io/cdap/plugin/google/common/GoogleAuthBaseConfig.java [hunk 5]
```diff
@@ -238,4 +298,53 @@ public String getClientSecret() {
   public String getRefreshToken() {
     return refreshToken;
   }
+
+  @Nullable
+  public String getServiceAccountFilePath() {
+    if (containsMacro(ACCOUNT_FILE_PATH)  || accountFilePath == null ||
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/google/common/GoogleAuthBaseConfig.java]: nit: replace accountFilePath == null || accountFilePath.isEmpty() -> Strings.insNullOrEmpty(accountFilePath)

File: src/main/java/io/cdap/plugin/google/common/GoogleAuthBaseConfig.java [hunk 6]
```diff
@@ -238,4 +298,53 @@ public String getClientSecret() {
...
+  }
+
+  @Nullable
+  public Boolean isServiceAccountFilePath() {
+    String serviceAccountType = getServiceAccountType();
+    return Strings.isNullOrEmpty(serviceAccountType) ? null : serviceAccountType.equals(SERVICE_ACCOUNT_FILE_PATH);
+  }
+
+  @Nullable
+  public String getServiceAccount() {
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/google/common/GoogleAuthBaseConfig.java]: This is only used in GoogleBasicAuthConfig.java where you're checking if the returned string is null or empty. The string value is unused. 

It seems this is unnecessary. You just need something like this:
```
if (config.isServiceAccountFilePath()) {
...
} else if (config. isServiceAccountJson()) {
...
} else {
// use default credentials
}
```

File: src/test/java/io/cdap/plugin/google/common/GoogleAuthBaseConfigTest.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+ */
+
+package io.cdap.plugin.google.common;
+
+import io.cdap.plugin.google.drive.source.GoogleDriveSourceConfig;
+import org.junit.Assert;
+import org.junit.Test;
+
+public class GoogleAuthBaseConfigTest {
+
```
Comment (rmstar) [inline src/test/java/io/cdap/plugin/google/common/GoogleAuthBaseConfigTest.java:28]: can you also add tests for validation error handling?

File: src/main/java/io/cdap/plugin/google/common/GoogleAuthBaseConfig.java (hunk 6 above)
Comment (flakrimjusufi) [reply src/main/java/io/cdap/plugin/google/common/GoogleAuthBaseConfig.java]: Removed in the last update.

File: src/test/java/io/cdap/plugin/google/common/GoogleAuthBaseConfigTest.java (hunk 7 above)
Comment (flakrimjusufi) [reply src/test/java/io/cdap/plugin/google/common/GoogleAuthBaseConfigTest.java:28]: Added tests for validation error handling in the last update.

**Comment by flakrimjusufi** [comment]
@rmstar 

In the last commit, GoogleSheets plugins are also updated with serviceAccountType as JSON, macros and documentation. 

Both Source and Sink were tested thoroughly and everything is working well.

File: src/main/java/io/cdap/plugin/google/common/GoogleAuthBaseConfig.java [hunk 8]
```diff
@@ -47,6 +51,11 @@
   public static final String REFRESH_TOKEN_LABEL = "Refresh token";
   public static final String ACCOUNT_FILE_PATH = "accountFilePath";
   public static final String DIRECTORY_IDENTIFIER = "directoryIdentifier";
+  public static final String NAME_SERVICE_ACCOUNT_TYPE = "serviceAccountType";
+  public static final String NAME_SERVICE_ACCOUNT_JSON = "serviceAccountJSON";
+  public static final String SERVICE_ACCOUNT_FILE_PATH = "filePath";
+  public static final String SERVICE_ACCOUNT_JSON = "JSON";
+  public static final String SCHEMA = "schema";
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/google/common/GoogleAuthBaseConfig.java:56]: where is this used?

File: src/main/java/io/cdap/plugin/google/sheets/source/GoogleSheetsSourceConfig.java [hunk 9]
```diff
@@ -228,16 +230,47 @@
...
+  }
+
   /**
    * Returns the instance of Schema.
    * @return The instance of Schema
    */
   public Schema getSchema() {
-    if (schema == null) {
-      if (dataSchemaInfo.isEmpty()) {
-        throw new RuntimeException("There are no headers to process. " +
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/google/sheets/source/GoogleSheetsSourceConfig.java:238]: Is this not relevant anymore?

File: src/main/java/io/cdap/plugin/google/sheets/source/GoogleSheetsSourceConfig.java [hunk 10]
```diff
@@ -228,16 +230,47 @@
...
-          "Perhaps no validation step was executed before schema generation.");
-      }
+    if (schema == null && !containsMacro("serviceAccountType")) {
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/google/sheets/source/GoogleSheetsSourceConfig.java]: What happens if service account type is not macro (set to default, i.e. file path), but the file path is a macro?

**Review by rmstar — COMMENTED** [review]
Can you add an ETL test with macros for google drive and sheets?

File: src/main/java/io/cdap/plugin/google/common/GoogleAuthBaseConfig.java (hunk 8 above)
Comment (flakrimjusufi) [reply src/main/java/io/cdap/plugin/google/common/GoogleAuthBaseConfig.java:56]: `SCHEMA` is used in deserialization of config: 

https://github.com/data-integrations/google-drive/blob/3074341092eecd2522ea5ce9ac1c40877e0fc03f/src/main/java/io/cdap/plugin/google/drive/source/GoogleDriveSourceConfig.java#L358-L360

File: src/main/java/io/cdap/plugin/google/sheets/source/GoogleSheetsSourceConfig.java (hunk 9 above)
Comment (flakrimjusufi) [reply src/main/java/io/cdap/plugin/google/sheets/source/GoogleSheetsSourceConfig.java:238]: It doesn't look relevant to me because we are already checking if Source folder (directory) contains any spreadsheets files in here: 

https://github.com/data-integrations/google-drive/blob/3074341092eecd2522ea5ce9ac1c40877e0fc03f/src/main/java/io/cdap/plugin/google/sheets/source/GoogleSheetsSourceConfig.java#L348-L354

...but I will revert this RuntimeException in getSchema().

File: src/main/java/io/cdap/plugin/google/sheets/source/GoogleSheetsSourceConfig.java (hunk 10 above)
Comment (flakrimjusufi) [reply src/main/java/io/cdap/plugin/google/sheets/source/GoogleSheetsSourceConfig.java]: Thanks for catching this. In this scenario the validation would fail. In the next update, I will also check if file path or service acccount json contains Macro.

**Comment by flakrimjusufi** [comment]
> Can you add an ETL test with macros for google drive and sheets?

I'm seeing lots of dependencies that are missing in `pom.xml`. Should I create a separate PR only with ETL Tests? It will take some time because first I will need to include all the missing dependencies, create the ETL Tests and then I will have to do a thorough test of both Google Drive and Google Sheets plugin.

File: src/main/java/io/cdap/plugin/google/sheets/source/GoogleSheetsSourceConfig.java (hunk 9 above)
Comment (rmstar) [reply src/main/java/io/cdap/plugin/google/sheets/source/GoogleSheetsSourceConfig.java:238]: hm why are we throwing a RuntimeException here? If config validation fails, we should add the failure to the failure collector (pass in FailureCollector, call collector.addFailure()).

**Comment by rmstar** [comment]
> Should I create a separate PR only with ETL Tests?

You can add a test in a separate PR if that's easier for you.

File: src/main/java/io/cdap/plugin/google/sheets/source/GoogleSheetsSourceConfig.java (hunk 9 above)
Comment (flakrimjusufi) [reply src/main/java/io/cdap/plugin/google/sheets/source/GoogleSheetsSourceConfig.java:238]: PR updated.

**Review by rmstar — APPROVED** [review]
please squash commits

**Comment by flakrimjusufi** [comment]
> please squash commits

Commits squashed.
